libs/@local/hash-isomorphic-utils/src/flows/types.ts

```diff
@@ -0,0 +1,163 @@
+import type { VersionedUrl } from "@blockprotocol/type-system";
+import type { Entity, EntityTypeWithMetadata } from "@local/hash-subgraph";
+
+import type { ProposedEntity } from "../ai-inference-types";
+
+/**
+ * Payload
+ */
+
+export type WebPage = {
+  url: string;
+  title: string;
+  textContent: string;
+};
+
+type PayloadKindValues = {
+  Text: string;
+  Number: number;
+  Boolean: boolean;
+  ProposedEntity: ProposedEntity;
+  Entity: Entity;
+  WebPage: WebPage;
+  EntityType: EntityTypeWithMetadata;
+  VersionedUrl: VersionedUrl;
+};
+
+type PayloadKind = keyof PayloadKindValues;
+
+type Payload = {
+  [K in keyof PayloadKindValues]: {
+    kind: K;
+    value: PayloadKindValues[K];
+  };
+}[keyof PayloadKindValues];
+
+/**
+ * Step Definition
+ */
+
+export type InputDefinition = {
+  name: string;
+  description?: string;
+  oneOfPayloadKinds: PayloadKind[];
+  array: boolean;
+  required: boolean;
+};
+
+export type OutputDefinition = {
+  name: string;
+  description?: string;
+  payloadKind: PayloadKind;
+  array: boolean;
+};
+
+export type TriggerDefinition = {
+  kind: "trigger";
+  name: string;
+  outputs?: OutputDefinition[];
+};
+
+export type ActionDefinition = {
+  kind: "action";
+  name: string;
+  inputs: InputDefinition[];
+  outputs: OutputDefinition[];
+};
+
+export type StepDefinition = ActionDefinition;
+
+/**
+ * Flow Definition
+ */
+
+type StepInputSource = {
+  inputName: string;
+} & (
+  | {
+      kind: "stepOutput";
+      sourceNodeId: string;
+      sourceNodeOutputName: string;
+      fallbackValue?: Payload;
+    }
+  /**
+   * An output provided by the trigger that started the flow
+   * (e.g. `userVisitedWebPageTrigger` provides a `visitedWebPAge` output)
+   */
+  | {
+      kind: "flow-trigger";
+      triggerOutputName: string;
+      fallbackValue?: Payload;
```

Comment:
I think we'll probably want triggers to have a `nodeId` (`stepId`?) because otherwise we'll need to special case them in various places when visualising the DAG... what're the advantages of treating them separately here rather than treating them as a node and using the object above?